plugin/path_checkouts.go

```diff
@@ -5,11 +5,349 @@ import (
+const secretAccessKeyType = "creds"
+
+func (b *backend) pathSetCheckOut() *framework.Path {
+	return &framework.Path{
+		Pattern: libraryPrefix + framework.GenericNameRegex("name") + "/check-out$",
+		Fields: map[string]*framework.FieldSchema{
+			"name": {
+				Type:        framework.TypeLowerCaseString,
+				Description: "Name of the set",
+				Required:    true,
+			},
+			"ttl": {
+				Type:        framework.TypeDurationSecond,
+				Description: "The length of time before the check-out will expire, in seconds.",
+			},
+		},
+		Operations: map[logical.Operation]framework.OperationHandler{
+			logical.ReadOperation: &framework.PathOperation{
+				Callback: b.operationSetCheckOut,
+				Summary:  "Check a service account out from the library.",
+			},
+		},
+		HelpSynopsis: `Check a service account out from the library.`,
+	}
+}
+
+func (b *backend) operationSetCheckOut(ctx context.Context, req *logical.Request, fieldData *framework.FieldData) (*logical.Response, error) {
+	setName := fieldData.Get("name").(string)
+
+	ttlPeriodRaw, ttlPeriodSent := fieldData.GetOk("ttl")
+	if !ttlPeriodSent {
+		ttlPeriodRaw = 0
+	}
+	requestedTTL := time.Duration(ttlPeriodRaw.(int)) * time.Second
+
+	set, err := readSet(ctx, req.Storage, setName)
+	if err != nil {
+		return nil, err
+	}
+	if set == nil {
+		return logical.ErrorResponse(`"%s" doesn't exist`, setName), nil
+	}
+
+	// Prepare the check-out we'd like to execute.
+	ttl := set.TTL
+	if ttlPeriodSent {
+		switch {
+		case set.TTL <= 0 && requestedTTL > 0:
+			// The set's TTL is infinite and the caller requested a finite TTL.
+			ttl = requestedTTL
+		case set.TTL > 0 && requestedTTL < set.TTL:
+			// The set's TTL isn't infinite and the caller requested a shorter TTL.
+			ttl = requestedTTL
+		}
+	}
+	newCheckOut := &CheckOut{
+		IsAvailable:         false,
+		BorrowerEntityID:    req.EntityID,
+		BorrowerClientToken: req.ClientToken,
+	}
+
+	// Check out the first service account available.
+	for _, serviceAccountName := range set.ServiceAccountNames {
+		// checkOut is its own function to avoid deferring the lock release inside a loop,
+		// which would cause a memory leak.
+		password, err := b.checkOut(ctx, req.Storage, serviceAccountName, newCheckOut)
+		if err == ErrCheckedOut {
+			continue
+		}
+		respData := map[string]interface{}{
+			"service_account_name": serviceAccountName,
+			"password":             password,
+		}
+		internalData := map[string]interface{}{
+			"service_account_name": serviceAccountName,
+		}
+		resp := b.Backend.Secret(secretAccessKeyType).Response(respData, internalData)
+		resp.Secret.Renewable = true
+		resp.Secret.TTL = ttl
+		resp.Secret.MaxTTL = set.MaxTTL
+		return resp, nil
+	}
+
+	// If we arrived here, it's because we never had a hit for a service account that was available.
+	resp, err := logical.RespondWithStatusCode(&logical.Response{
+		Warnings: []string{"No service accounts available for check-out, please try again later."},
+	}, req, 429)
```

Comment:
Why is error code 429 (Too Many Requests) used? That seems to be focused on indicating that the user is being rate limited. Perhaps [409 (Conflict)](https://tools.ietf.org/html/rfc7231#section-6.5.8) would be a better fit?

Reply:
This was determined as part of the RFC. It was originally a different status code, 423 locked, but that's a WebDav status code. @jefferai suggested 429 Too Many Requests because it essentially _is_ too many requests. However, I can see the case for 409 as well. 

If we do 429, clients may typically react by using exponential backoff and retrying, whereas their behavior with 409 is not as defined. My guess is that most clients that receive a 4xx they haven't added special handling for will just fail assuming it's caller error.

Reply:
Yeah there isn't a great fitting code, as you well know. I'm probably less for 409 than I am against 429. To me the intent of [429](https://tools.ietf.org/html/rfc6585#section-4) is pretty specific. 

> The 429 status code indicates that the user has sent too many requests in a given amount of time ("rate limiting").

Returning that when the user has perhaps sent one request ever seems like a misuse, and prompts the question of what is the rate limit.

This isn't a blocker because it is trivial to change later (e.g. before GA), but I would like to perhaps reengage Jeff or Brian and see if they're amenable to something different.

Reply:
Added it to our next ecosystem sync convo.